### PR TITLE
[SPARK-21232][SQL][SparkR][PYSPARK] New built-in SQL function - Data_Type

### PR DESCRIPTION
**Pull request by mmolimar**
## What changes were proposed in this pull request?

New built-in function to get the data type of columns in SQL.

## How was this patch tested?

Unit tests included.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 1]
```diff
@@ -209,6 +209,18 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
       Row(2743272264L, 2180413220L))
   }
 
+  test("misc data_type function") {
+    val df = Seq(("a", false)).toDF("a", "b")
+
+    checkAnswer(
+      df.select(data_type($"a"), data_type($"b")),
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: I think we can easily get the types via

```scala
val df = Seq(("a", false)).toDF("a", "b")
df.schema.map(_.dataType.simpleString)
```

I just wonder in which case we need these type per each row in the dataframe.

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -1248,6 +1248,17 @@ def hash(*cols):
     return Column(jc)
 
 
+def data_type(col):
+    """Returns the data type of the given column.
+
+    >>> spark.createDataFrame([('ABC',)], ['a']).select(data_type('a').alias('data_type')).collect()
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: In Python, I think we could do this as below:

```python
>>> df = spark.createDataFrame([('ABC',)], ['a'])
>>> [s.dataType.simpleString() for s in df.schema]
['string']
```

File: R/pkg/R/functions.R [hunk 3]
```diff
@@ -598,6 +598,25 @@ setMethod("hash",
...
+#' Returns the data type of a given column.
+#'
+#' @param x Column to get the data type.
+#'
+#' @rdname data_type
+#' @name data_type
+#' @family misc functions
+#' @aliases data_type,Column-method
+#' @export
+#' @examples \dontrun{data_type(df$c)}
```
Comment (HyukjinKwon) [inline R/pkg/R/functions.R]: In R, I think we could do this as below:

```r
> df <- createDataFrame(iris)
> lapply(schema(df)$fields(), function(s) { s$dataType.simpleString() })
[[1]]
[1] "double"

[[2]]
[1] "double"

[[3]]
[1] "double"

[[4]]
[1] "double"

[[5]]
[1] "string"
```

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala (hunk 1 above)
Comment (mmolimar) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: The idea would be to know the type based on the value itself, not by the schema (i.e. the value could be null):
```scala
val df = spark.sparkContext.parallelize(
      StringData(null) ::
      StringData("a") :: Nil).toDF()
df.select(data_type(col("s"))) //you get null and string in this case
df.schema.map(_.dataType.simpleString) // you just get string
```
On the other hand, it'd be nice to have this SQL function as we do in some databases.

Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: For me, it sounds the point is to know where it is `null`. I think it'd be more persuasive if you leave some links here with other equivalent SQL function in other databases and match the behaviour.

Comment (mmolimar) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: I can't think of a SQL db which queries the datatype without using the table schema. However, for example in MongoDB, you can get something like that using the $type operator or 'typeof'.

Comment (mmolimar) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: @HyukjinKwon so?

Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Not sure. I think we need a reference for a function reference in an RDBMS (SQL to be more correct).

**Comment by gatorsmile** [comment]
Which systems have such function calls? Could you give some examples?

**Comment by mmolimar** [comment]
In some SQL db you have to query explicitly the table schema, ie: ``select data_type from all_tab_columns where table_name = 'my_table'``or something like that.
In case of the ARQ engine from Apache Jena you can call this function in SPARQL (see [W3C-SPARQL](https://www.w3.org/TR/rdf-sparql-query/#func-datatype).
I find it useful in order to avoid to query the schema.

**Comment by felixcheung** [comment]
ok to test

**Comment by felixcheung** [comment]
can you rebase? (there is a conflict)

File: R/pkg/R/functions.R [hunk 4]
```diff
@@ -653,6 +653,25 @@ setMethod("hash",
...
 
+#' data_type
+#'
+#' Returns the data type of a given column.
+#'
+#' @param x Column to get the data type.
+#'
+#' @rdname data_type
+#' @name data_type
+#' @family misc functions
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: can you follow the pattern, like for "hash" above? we don't have individual rdname etc for each function

**Review by felixcheung — COMMENTED** [review]
comment on the R changes.
also this is incomplete - you need to add to the NAMESPACE file otherwise the function is not public and also should always have tests.

**Comment by mmolimar** [comment]
@felixcheung Everything done!

File: R/pkg/R/functions.R [hunk 5]
```diff
@@ -679,6 +679,19 @@ setMethod("hash",
             column(jc)
           })
 
+#' @details
+#' \code{data_type}: Returns the data type of a given column.
+#'
+#' @rdname column_misc_functions
+#' @aliases data_type data_type,Column-method
+#' @examples \dontrun{data_type(df$c)}
+setMethod("data_type",
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: add `@note` like example above

File: R/pkg/R/functions.R [hunk 6]
```diff
@@ -679,6 +679,19 @@ setMethod("hash",
             column(jc)
           })
 
+#' @details
+#' \code{data_type}: Returns the data type of a given column.
+#'
+#' @rdname column_misc_functions
+#' @aliases data_type data_type,Column-method
+#' @examples \dontrun{data_type(df$c)}
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: example should no be added here - see the example for `hash`

File: R/pkg/NAMESPACE [hunk 7]
```diff
@@ -236,6 +236,7 @@ exportMethods("%<=>%",
               "current_date",
               "current_timestamp",
               "hash",
+              "data_type",
```
Comment (felixcheung) [inline R/pkg/NAMESPACE]: I know this list isn't completely sorted, let's sort this?
you can move "hash" and "data_type"

**Review by felixcheung — COMMENTED** [review]
as mention, you need to add tests for R

File: R/pkg/R/functions.R (hunk 5 above)
Comment (felixcheung) [reply R/pkg/R/functions.R]: 2.4.0

File: R/pkg/R/functions.R (hunk 6 above)
Comment (felixcheung) [reply R/pkg/R/functions.R]: see this line of the code example of hash
https://github.com/mmolimar/spark/blob/ed52e2f856f78fb2dca23b6be2f682caa0a88c81/R/pkg/R/functions.R#L176

File: R/pkg/R/functions.R [hunk 8]
```diff
@@ -679,6 +679,19 @@ setMethod("hash",
             column(jc)
           })
 
+#' @details
+#' \code{data_type}: Returns the data type of a given column.
+#'
+#' @rdname column_misc_functions
+#' @aliases data_type data_type,Column-method
+#' @note data_type since 2.3.0
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: 2.4.0

**Review by felixcheung — COMMENTED** [review]
code LGTM, did we address https://github.com/apache/spark/pull/18447/files#r130032578 ?

**Comment by mmolimar** [comment]
so @felixcheung ?

**Comment by felixcheung** [comment]
I guess we still have an open question on https://github.com/apache/spark/pull/18447/files#r130032578

**Comment by HyukjinKwon** [comment]
so @mmolimar ?

**Comment by mmolimar** [comment]
Hi @HyukjinKwon 
For me it's fine:
"In some SQL db you have to query explicitly the table schema, ie: select data_type from all_tab_columns where table_name = 'my_table'or something like that.
In case of the ARQ engine from Apache Jena you can call this function in SPARQL (see [W3C-SPARQL](https://www.w3.org/TR/rdf-sparql-query/#func-datatype)).
I find it useful in order to avoid to query the schema."

File: R/pkg/R/generics.R [hunk 9]
```diff
@@ -950,6 +949,10 @@ setGeneric("grouping_bit", function(x) { standardGeneric("grouping_bit") })
 #' @name NULL
 setGeneric("grouping_id", function(x, ...) { standardGeneric("grouping_id") })
 
+#' @rdname column_misc_functions
+#' @name NULL
+setGeneric("hash", function(x, ...) { standardGeneric("hash") })
```
Comment (HyukjinKwon) [inline R/pkg/R/generics.R]: I would avoid unrelated changes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 10]
```diff
@@ -98,6 +98,29 @@ case class AssertTrue(child: Expression) extends UnaryExpression with ImplicitCa
...
+      > SELECT _FUNC_("a");
+       string
+      > SELECT _FUNC_(OL);
+       bigint
+  """)
+case class GetDataType(child: Expression) extends UnaryExpression {
+
+  override def dataType: DataType = StringType
+
+  override def nullSafeEval(input: Any): Any = UTF8String.fromString(child.dataType.simpleString)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: It should be `catalogString` instead of `simpleString`.

**Comment by HyukjinKwon** [comment]
@MaxGekk and @rxin WDYT on this function? I am kind of neutral on this and wonder if you guys have some preferences.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 11]
```diff
@@ -98,6 +98,29 @@ case class AssertTrue(child: Expression) extends UnaryExpression with ImplicitCa
...
 
+@ExpressionDescription(
+  usage = "_FUNC_(expr) - Returns the data type of the `expr`.",
+  extended = """
+    Examples:
+      > SELECT _FUNC_("a");
+       string
+      > SELECT _FUNC_(OL);
+       bigint
+  """)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: `since` should be added

**Comment by MaxGekk** [comment]
@HyukjinKwon I also neutral since I don't have potential use cases for the function.

**Comment by HyukjinKwon** [comment]
@MaxGekk, I was thinking we could use this instead of `schema_of` somehow

**Comment by HyukjinKwon** [comment]
@mmolimar, let's leave this closed if there's no explicit positive feedback within few days.

**Comment by rxin** [comment]
Yea I'd probably reject this for now, until we see bigger needs for it.

**Comment by HyukjinKwon** [comment]
ping @mmolimar to close
